### PR TITLE
🍒Refactor dms loader

### PR DESCRIPTION
**Pull request by doctrino**
# Description

Pure refactoring

The `DMSLoader` is currently implicitly creating instance spaces. This PR moves out the code for creating instance spaces into `InstanceSpaceLoader`. 

## Bump

- [ ] Patch
- [ ] Minor
- [x] Skip

### CODE REVIEW

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+                    DataModelsAcl.Action.Read,
+                ],
+                scope=DataModelsAcl.Scope.All(),
+            )
+        ]
+
+    def _upload_to_cdf(
+        self,
+        client: NeatClient,
+        items: list[dm.SpaceApply],
```
Comment (nikokaoja) [inline cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:85]:
```suggestion
        spaces: list[dm.SpaceApply],
```

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        dry_run: bool,
+        read_issues: IssueList,
+        class_name: str | None = None,
+    ) -> Iterable[UploadResult]:
+        cdf_items = client.data_modeling.spaces.retrieve([item.space for item in items])
```
Comment (nikokaoja) [inline cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:90]:
```suggestion
        cdf_spaces = client.data_modeling.spaces.retrieve([item.space for item in items])
```

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        cdf_idem_by_id = {item.space: item for item in cdf_items}
```
Comment (nikokaoja) [inline cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:91]:
```suggestion
        cdf_space_by_id = {item.space: item for item in cdf_items}
```

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    def _load(
+        self, stop_on_exception: bool = False
+    ) -> Iterable[dm.SpaceApply | NeatIssue | type[_END_OF_CLASS] | _START_OF_CLASS]:
+        yield from self._lookup_issues
+        seen: set[str] = set()
+        for space_str in self.space_by_instance_uri.values():
+            if space_str in seen or space_str in COGNITE_SPACES:
+                continue
+            yield dm.SpaceApply(space=space_str)
+            seen.add(space_str)
```
Comment (nikokaoja) [inline cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:163]: If you have very large number of instances maybe it is quicker to execute set operation over values

```suggestion
        for space_str in set(self.space_by_instance_uri.values()):
            if space_str in seen or space_str in COGNITE_SPACES:
                continue
            yield dm.SpaceApply(space=space_str)
            seen.add(space_str)
```

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+            # Adding a dummy entry to ensure that the instance space is included
+            self._space_by_instance_uri[URIRef(self.instance_space)] = self.instance_space
+            return
+        if self.graph_store is None:
+            raise ValueError("Graph store must be provided to lookup spaces")
+        # Case 2: Use the source space, i.e., the space of the instances when extracted from CDF
+        if self.use_source_space:
+            self._lookup_instance_uris(self.graph_store)
+        # Case 3: Use a property on each instance to determine the space.
+        elif self.space_property is not None:
```
Comment (nikokaoja) [inline cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:178]: Wouldn't this be 

```
if self.use_source_space and self.space_property is None: 

```

?

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+            if self.instance_space is None:
+                raise ValueError(
+                    f"Missing fallback instance space. This is required when using '{self.space_property=}'"
+                )
+            self._space_by_instance_uri = defaultdict(lambda: cast(str, self.instance_space))
+            self._lookup_space_property(self.graph_store, self.space_property)
+        else:
+            raise ValueError("Either 'instance_space", "space_property', or 'use_source_space' must be provided. ")
+
+    def _lookup_instance_uris(self, graph_store: NeatInstanceStore) -> None:
```
Comment (nikokaoja) [inline cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:188]:
```suggestion
    def _lookup_space_via_instance_uris(self, graph_store: NeatInstanceStore) -> None:
```

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+        for instance_uri in instance_iterable:
+            namespace, external_id = split_uri(instance_uri)
+            space = namespace_as_space(namespace)
+            if space is None:
+                error = ResourceCreationError(instance_uri, "Unknown", f"Could not find space for {instance_uri!s}.")
+                self._lookup_issues.append(error)
+            else:
+                self._space_by_instance_uri[instance_uri] = space
+
+    def _lookup_space_property(self, graph_store: NeatInstanceStore, space_property: str) -> None:
```
Comment (nikokaoja) [inline cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:203]:
```suggestion
    def _lookup_space_via_property(self, graph_store: NeatInstanceStore, space_property: str) -> None:
```

File: cognite/neat/core/_instances/queries/_select.py [hunk 8]
```diff
@@ -120,6 +120,18 @@ def list_instances_ids(
...
+    def list_instance_object_ids(self, limit: int = -1, named_graph: URIRef | None = None) -> Iterable[URIRef]:
+        query = (
+            "SELECT DISTINCT ?object WHERE { "
+            "?subject ?predicate ?object . FILTER(?predicate != rdf:type && isURI(?object))"
+            "}"
+        )
+        if limit != -1:
+            query += f" LIMIT {limit}"
+
+        # MyPy is not very happy with RDFLib, so just ignore the type hinting here
```
Comment (nikokaoja) [inline cognite/neat/core/_instances/queries/_select.py:132]: I think we should propose changes to RDFLib to be more MyPy compliant

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+            self._space_by_instance_uri = defaultdict(lambda: cast(str, self.instance_space))
+            self._lookup_space_property(self.graph_store, self.space_property)
+        else:
+            raise ValueError("Either 'instance_space", "space_property', or 'use_source_space' must be provided. ")
+
+    def _lookup_instance_uris(self, graph_store: NeatInstanceStore) -> None:
+        instance_iterable = itertools.chain(
+            (res[0] for res in graph_store.queries.select.list_instances_ids()),
+            graph_store.queries.select.list_instance_object_ids(),
+        )
```
Comment (nikokaoja) [inline cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:192]: why two calls ? isnt `list_instances_ids` enough ?

**Review by nikokaoja — CHANGES_REQUESTED** [review]

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py (hunk 9 above)
Comment (doctrino) [reply cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:192]: No, there can be instance ids both in the subject and object.

File: cognite/neat/core/_instances/queries/_select.py (hunk 8 above)
Comment (doctrino) [reply cognite/neat/core/_instances/queries/_select.py:132]: No sure it is possible as it is in the translation layer between the SPARQL strings and Python

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py (hunk 5 above)
Comment (doctrino) [reply cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:178]: No, we have three ways of setting space. Either you just provide it, it set based on a Literal, or it is set based on the space used when reading from CDF. This is setting from a space literal.

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py (hunk 9 above)
Comment (nikokaoja) [reply cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:192]: true, as we allow for "open world" assumption (where object is not defined in current graph)

**Review by nikokaoja — APPROVED** [review]

**Review by nikokaoja — DISMISSED** [review]

File: tests/tests_unit/test_graph/test_loaders/test_rdf_to_instance_space.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+from cognite.neat.core._constants import DEFAULT_SPACE_URI
+from cognite.neat.core._instances.extractors import AssetsExtractor, BaseExtractor
+from cognite.neat.core._instances.loaders import InstanceSpaceLoader
+from cognite.neat.core._shared import Triple
+from cognite.neat.core._store import NeatInstanceStore
+from cognite.neat.core._utils.upload import UploadResult
+from tests.data import InstanceData
+
+
+def load_instance_spaces_test_cases():
```
Comment (rsjr) [inline tests/tests_unit/test_graph/test_loaders/test_rdf_to_instance_space.py:17]: Can you add a test case to cover when graph_store is None? What is the exception/logging/NeatIssue we are sending to the user/session when the whole thing is None.

**Review by rsjr — COMMENTED** [review]
Some comments

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+        instance_space: str | None = None,
+        space_property: str | None = None,
+        use_source_space: bool = False,
+        neat_prefix_by_predicate_uri: dict[URIRef, str] | None = None,
+    ) -> None:
+        self.graph_store = graph_store
+        self.instance_space = instance_space
+        self.space_property = space_property
+        self.use_source_space = use_source_space
+        self.neat_prefix_by_predicate_uri = neat_prefix_by_predicate_uri or {}
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:56]: No, it is a union of dict and None, this is common pattern to ensure that the attribute `self.neat_prefix_by_predicate_uri` is a dict.

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+        if self.graph_store is None:
+            raise ValueError("Graph store must be provided to lookup spaces")
+        # Case 2: Use the source space, i.e., the space of the instances when extracted from CDF
+        if self.use_source_space:
+            self._lookup_space_via_instance_uris(self.graph_store)
+        # Case 3: Use a property on each instance to determine the space.
+        elif self.space_property is not None:
+            if self.instance_space is None:
+                raise ValueError(
+                    f"Missing fallback instance space. This is required when using '{self.space_property=}'"
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:181]: No, but I can make it explicit.

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+                )
+            self._space_by_instance_uri = defaultdict(lambda: cast(str, self.instance_space))
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:183]: Correct.

File: cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+        self.graph_store = graph_store
+        self.instance_space = instance_space
+        self.space_property = space_property
+        self.use_source_space = use_source_space
+        self.neat_prefix_by_predicate_uri = neat_prefix_by_predicate_uri or {}
+
+        self._lookup_issues = IssueList()
+
+        self._has_looked_up = False
+        self._space_by_instance_uri: dict[URIRef, str] = {}
```
[the remark this answers is not in the document]
Comment (doctrino) [reply cognite/neat/core/_instances/loaders/_rdf_to_instance_space.py:61]: defaultdict is a dict. So not sure I agree that this is funky. Adding a comment on why.

**Review by rsjr — COMMENTED** [review]
risk review ok

**Review by nikokaoja — APPROVED** [review]
